diff -Nuar linux-3.1.10/arch/arm/mach-davinci/board-omapl138-hawk.c linux-3.1.10_camera/arch/arm/mach-davinci/board-omapl138-hawk.c
--- linux-3.1.10/arch/arm/mach-davinci/board-omapl138-hawk.c	2012-05-11 17:50:15.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/board-omapl138-hawk.c	2014-03-22 14:30:15.467095513 +0400
@@ -15,6 +15,8 @@
 #include <linux/console.h>
 #include <linux/gpio.h>
 #include <linux/i2c.h>
+#include <linux/delay.h>
+#include <media/soc_camera.h>
 
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
@@ -27,6 +29,8 @@
 #define DA850_LCD_PWR_PIN		GPIO_TO_PIN(2, 8)
 #define DA850_LCD_BL_PIN		GPIO_TO_PIN(2, 15)
 
+#define CAMERA_RESET_GPIO   GPIO_TO_PIN(2, 10)
+
 #ifdef  CONFIG_HAWKBOARD_LCDK
 #define DA850_HAWK_MMCSD_CD_PIN		GPIO_TO_PIN(4, 0)
 #else
@@ -335,10 +339,21 @@
 static __init int omapl138_lcdk_init_cpufreq(void) { return 0; }
 #endif
 
+#if defined(CONFIG_SOC_CAMERA_OV5642) ||\
+		defined(CONFIG_SOC_CAMERA_OV5642_MODULE)
+static struct soc_camera_link link_ov5642;
+
 /* I2C */
 static struct i2c_board_info __initdata omapl138_hawk_i2c_devices[] = {
 	{
-		I2C_BOARD_INFO("tlv320aic3x", 0x18),
+		I2C_BOARD_INFO("ov5642", 0x3c),
+		.platform_data = &link_ov5642,
+	},
+#else
+static struct i2c_board_info __initdata omapl138_hawk_i2c_devices[] = {
+#endif
+	{
+		I2C_BOARD_INFO("tlv320aic3x", 0x1A),
 	},
 };
 
@@ -357,12 +372,59 @@
 
 	ret = da8xx_register_i2c(0, &omapl138_hawk_i2c_0_pdata);
 	if (ret)
-		pr_warning("omapl138_hawk_init: i2c0 registration failed: %d\n",
-				ret);
+		pr_warning("omapl138_hawk_init: i2c0 registration failed: %d\n", ret);
+
 	i2c_register_board_info(1, omapl138_hawk_i2c_devices,
 			ARRAY_SIZE(omapl138_hawk_i2c_devices));
 }
 
+/*
+ * soc camera
+ */
+#if defined(CONFIG_SOC_CAMERA_OV5642) ||\
+	  defined(CONFIG_SOC_CAMERA_OV5642_MODULE)
+
+static int i2c_camera_power(struct device * dev, int on) {
+	/* enable or disable the camera */
+	gpio_set_value(CAMERA_RESET_GPIO, !on);
+
+	if (!on) {
+		return 0;
+	}
+
+	/* if enabled, give a reset impulse */
+	gpio_set_value(CAMERA_RESET_GPIO, 1);
+	msleep(20);
+
+	gpio_set_value(CAMERA_RESET_GPIO, 0);
+	msleep(100);
+
+	return 0;
+}
+
+static unsigned long camera_query_bus_param(struct soc_camera_link *link) {
+	/* FIXME: need to be checked and replaced with a correct value */
+	return SOCAM_DATAWIDTH_16;
+}
+
+static struct soc_camera_link link_ov5642 = {
+	.bus_id = 0,
+	.board_info = &omapl138_hawk_i2c_devices[0],
+	.module_name = "ov5642",
+	.i2c_adapter_id = 1,
+	.power = i2c_camera_power,
+	.query_bus_param = camera_query_bus_param,
+};
+
+static struct platform_device camera_ov5642 = {
+	.name = "soc-camera-pdrv",
+	.id = 0,
+	.dev = {
+		.platform_data = &link_ov5642,
+	},
+};
+#endif
+
 /* VGA */
 static const short omapl138_hawk_lcdc_pins[] = {
 	DA850_GPIO2_8, DA850_GPIO2_15,
@@ -465,6 +527,47 @@
 	da8xx_register_mcasp(0, &omapl138_hawk_snd_data);
 }
 
+/*
+ *static const short omapl138_hawk_pwm_pins[] __initconst = {
+ *  DA850_ECAP0_APWM0, DA850_ECAP1_APWM1, DA850_EPWM0B, DA850_EPWM0A,
+ *  DA850_EPWMSYNCI, DA850_EPWMSYNC0, DA850_ECAP2_APWM2, DA850_EHRPWMGLUETZ,
+ *  DA850_EPWM2B, DA850_EPWM2A, DA850_EPWM1B, DA850_EPWM1A,
+ *  -1
+ *}
+ *
+ *static void omapl138_pwm_init(void)
+ *{
+ *  int ret = davinci_cfg_reg_list(omapl138_pwm_pins);
+ *}
+ */
+
+static const short omapl138_camera_pins[] __initconst = {
+	DA850_CLKOUT,		/* clock pin */
+	DA850_GPIO2_10,	/* reset pin */
+	DA850_VPIF_DIN1,  DA850_VPIF_DIN2,  DA850_VPIF_DIN3,  DA850_VPIF_DIN4,
+	DA850_VPIF_DIN5,  DA850_VPIF_DIN6,  DA850_VPIF_DIN7,  DA850_VPIF_DIN8,
+	DA850_VPIF_DIN9,  DA850_VPIF_DIN10, DA850_VPIF_DIN11, DA850_VPIF_DIN12,
+	DA850_VPIF_DIN13_FIELD, DA850_VPIF_DIN14_HSYNC, DA850_VPIF_DIN15_VSYNC,
+	-1
+};
+
+static void omapl138_camera_init(void) {
+	int err = 0;
+
+	davinci_cfg_reg_list(omapl138_camera_pins);
+
+	if (0 != (err = gpio_request(CAMERA_RESET_GPIO, "GpioTest"))) {
+		pr_alert("TestGpio driver: gpio reuest failed! Return status: %d\n", err);
+		return;
+	}
+
+	if (0 != (err = gpio_direction_output(CAMERA_RESET_GPIO, 0))) {
+		pr_alert("TestGpio driver: cannot setup direction! Return status: %d\n", err);
+		gpio_free(CAMERA_RESET_GPIO);
+		return;
+	}
+}
+
 static __init void omapl138_hawk_init(void)
 {
 	int ret;
@@ -509,9 +612,12 @@
 		pr_warning("omapl138_hawk_init: cpuidle registration failed: %d\n",
 				ret);
 
+	platform_device_register(&camera_ov5642);
 	omapl138_hawk_i2c_init();
 	omapl138_hawk_display_init();
 	omapl138_hawk_sound_init();
+	omapl138_camera_init();
+	omapl138_clkout_init();
 
 	ret = da850_register_sata(HAWKBOARD_SATA_REFCLKPN_RATE);
 	if (ret)
@@ -533,6 +639,7 @@
 static void __init omapl138_hawk_map_io(void)
 {
 	da850_init();
+	/*davinci_set_pllrate();*/
 }
 
 #ifdef CONFIG_HAWKBOARD_LCDK
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/clock.c linux-3.1.10_camera/arch/arm/mach-davinci/clock.c
--- linux-3.1.10/arch/arm/mach-davinci/clock.c	2012-01-18 19:33:18.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/clock.c	2014-03-21 14:28:10.413493403 +0400
@@ -25,6 +25,7 @@
 #include <mach/clock.h>
 #include <mach/psc.h>
 #include <mach/cputype.h>
+#include <mach/da8xx.h>
 #include "clock.h"
 
 static LIST_HEAD(clocks);
@@ -346,6 +347,8 @@
 	v |= ratio | PLLDIV_EN;
 	__raw_writel(v, pll->base + clk->div_reg);
 
+	omapl138_clkout_init();
+
 	v = __raw_readl(pll->base + PLLCMD);
 	v |= PLLCMD_GOSET;
 	__raw_writel(v, pll->base + PLLCMD);
@@ -492,6 +495,7 @@
 	if (pll->flags & PLL_HAS_POSTDIV)
 		__raw_writel(postdiv, pll->base + POSTDIV);
 
+	omapl138_clkout_init();
 	udelay(PLL_RESET_TIME);
 
 	/* Bring PLL out of reset */
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/clock.h linux-3.1.10_camera/arch/arm/mach-davinci/clock.h
--- linux-3.1.10/arch/arm/mach-davinci/clock.h	2012-01-18 19:33:18.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/clock.h	2014-03-21 14:28:15.633539462 +0400
@@ -25,6 +25,18 @@
 #define PLLCTL_PLLENSRC	BIT(5)
 #define PLLCTL_CLKMODE  BIT(8)
 
+#define PLLOCSEL				0x104
+#define PLLOCSEL0_OSCIN  0x14
+#define PLLOCSEL0_CLK1   0x17
+#define PLLOCSEL0_CLK2   0x18
+#define PLLOCSEL0_CLK3   0x19
+#define PLLOCSEL0_CLK4   0x1A
+#define PLLOCSEL0_CLK5   0x1B
+#define PLLOCSEL0_CLK6   0x1C
+#define PLLOCSEL0_CLK7   0x1D
+#define PLLOCSEL0_OBSCLK1 0x1E
+#define PLLOCSEL0_DISABLE 0x1F
+
 #define PLLM		0x110
 #define PLLM_PLLM_MASK  0xff
 
@@ -32,6 +44,7 @@
 #define PLLDIV1         0x118
 #define PLLDIV2         0x11c
 #define PLLDIV3         0x120
+#define OSCDIV          0x124
 #define POSTDIV         0x128
 #define BPDIV           0x12c
 #define PLLCMD		0x138
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/da850.c linux-3.1.10_camera/arch/arm/mach-davinci/da850.c
--- linux-3.1.10/arch/arm/mach-davinci/da850.c	2012-01-18 19:33:18.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/da850.c	2014-03-21 14:28:10.413493403 +0400
@@ -52,6 +52,7 @@
 static struct pll_data pll0_data = {
 	.num		= 1,
 	.phys_base	= DA8XX_PLL0_BASE,
+	.base  = 0,
 	.flags		= PLL_HAS_PREDIV | PLL_HAS_POSTDIV,
 };
 
@@ -122,6 +123,7 @@
 static struct pll_data pll1_data = {
 	.num		= 2,
 	.phys_base	= DA850_PLL1_BASE,
+	.base = 0,
 	.flags		= PLL_HAS_POSTDIV,
 };
 
@@ -148,38 +150,9 @@
 static struct clk pll1_sysclk3 = {
 	.name		= "pll1_sysclk3",
 	.parent		= &pll1_clk,
-	.flags		= CLK_PLL,
 	.div_reg	= PLLDIV3,
 };
 
-static struct clk pll1_sysclk4 = {
-	.name		= "pll1_sysclk4",
-	.parent		= &pll1_clk,
-	.flags		= CLK_PLL,
-	.div_reg	= PLLDIV4,
-};
-
-static struct clk pll1_sysclk5 = {
-	.name		= "pll1_sysclk5",
-	.parent		= &pll1_clk,
-	.flags		= CLK_PLL,
-	.div_reg	= PLLDIV5,
-};
-
-static struct clk pll1_sysclk6 = {
-	.name		= "pll0_sysclk6",
-	.parent		= &pll0_clk,
-	.flags		= CLK_PLL,
-	.div_reg	= PLLDIV6,
-};
-
-static struct clk pll1_sysclk7 = {
-	.name		= "pll1_sysclk7",
-	.parent		= &pll1_clk,
-	.flags		= CLK_PLL,
-	.div_reg	= PLLDIV7,
-};
-
 static struct clk i2c0_clk = {
 	.name		= "i2c0",
 	.parent		= &pll0_aux_clk,
@@ -396,10 +369,6 @@
 	CLK(NULL,		"pll1_aux",	&pll1_aux_clk),
 	CLK(NULL,		"pll1_sysclk2",	&pll1_sysclk2),
 	CLK(NULL,		"pll1_sysclk3",	&pll1_sysclk3),
-	CLK(NULL,		"pll1_sysclk4",	&pll1_sysclk4),
-	CLK(NULL,		"pll1_sysclk5",	&pll1_sysclk5),
-	CLK(NULL,		"pll1_sysclk6",	&pll1_sysclk6),
-	CLK(NULL,		"pll1_sysclk7",	&pll1_sysclk7),
 	CLK("i2c_davinci.1",	NULL,		&i2c0_clk),
 	CLK(NULL,		"timer0",	&timerp64_0_clk),
 	CLK("watchdog",		NULL,		&timerp64_1_clk),
@@ -582,7 +551,7 @@
 	MUX_CFG(DA850, EMA_D_15,	8,	0,	15,	1,	false)
 	MUX_CFG(DA850, EMA_BA_1,	5,	24,	15,	1,	false)
 	MUX_CFG(DA850, EMA_CLK,		6,	0,	15,	1,	false)
-	MUX_CFG(DA850, EMA_WAIT_1,	6,	24,	15,	1,	false)
+	MUX_CFG(DA850, EMA_WAIT_1,6,	24,	15,	1,  false)
 	MUX_CFG(DA850, NEMA_CS_2,	7,	0,	15,	1,	false)
 	/* GPIO function */
 	MUX_CFG(DA850, GPIO2_4,		6,	12,	15,	8,	false)
@@ -595,6 +564,27 @@
 	MUX_CFG(DA850, GPIO4_1,		10,	24,	15,	8,	false)
 	MUX_CFG(DA850, GPIO6_13,	13,	8,	15,	8,	false)
 	MUX_CFG(DA850, RTC_ALARM,	0,	28,	15,	2,	false)
+	/* CLKOUT function (used as clock for camera, 24MHz) */
+	MUX_CFG(DA850, CLKOUT,    13,  4, 15, 1,  false)
+	/* RESET function (used to reset camera)*/
+	MUX_CFG(DA850, GPIO2_10,  5,  20, 15, 8,  false)
+	/* Video port configuration */
+	MUX_CFG(DA850, VPIF_DIN0, 15, 4,  15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN1, 15, 0,  15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN2, 14, 28, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN3, 14, 24, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN4, 14, 20, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN5, 14, 16, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN6, 14, 12, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN7, 14,  8, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN8, 16,  4, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN9, 16,  0, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN10, 15, 28, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN11, 15, 24, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN12, 15, 20, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN13_FIELD, 15, 16, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN14_HSYNC, 15, 12, 15, 1, false)
+	MUX_CFG(DA850, VPIF_DIN15_VSYNC, 15, 8, 15, 1, false)
 #endif
 };
 
@@ -1147,3 +1137,65 @@
 	v &= ~CFGCHIP3_PLL1_MASTER_LOCK;
 	__raw_writel(v, DA8XX_SYSCFG0_VIRT(DA8XX_CFGCHIP3_REG));
 }
+
+void __init omapl138_clkout_init(void) {
+	unsigned int v;
+
+
+	/* setup CLKOUT output */
+	if ( pll0_clk.pll_data->base != NULL ) {
+		pr_warning("Setup pll0...\n");
+		v = __raw_readl(pll0_clk.pll_data->base + PLLOCSEL);
+		pr_warning("Before write pll0 %s: 0x%x\n", "PLLOCSEL", v);
+		v &= ~0xff;
+		v |= PLLOCSEL0_OSCIN;
+		__raw_writel(v, pll0_clk.pll_data->base + PLLOCSEL);
+		v = __raw_readl(pll0_clk.pll_data->base + PLLOCSEL);
+		pr_warning("After write pll0 %s: 0x%x\n", "PLLOCSEL", v);
+
+		v = __raw_readl(pll0_clk.pll_data->base + OSCDIV);
+		pr_warning("Before write pll0 %s: 0x%x\n", "OSCDIV", v);
+		v &= ~0xffff;
+		v |= 0x8000;
+		__raw_writel(v, pll0_clk.pll_data->base + OSCDIV);
+		v = __raw_readl(pll0_clk.pll_data->base + OSCDIV);
+		pr_warning("After write pll0 %s: 0x%x\n", "OSCDIV", v);
+
+	} else {
+		pr_warning("Cannot setup output clock for pll0!\n");
+	}
+
+/*
+ *  if ( pll1_clk.pll_data->base != NULL ) {
+ *    pr_warning("Setup pll1...\n");
+ *
+ *    v = __raw_readl(pll1_clk.pll_data->base + PLLOCSEL);
+ *    pr_warning("Before write %s: 0x%x\n", "PLLOCSEL", v);
+ *    v &= ~0xff;
+ *    v |= PLLOCSEL0_OSCIN;
+ *    __raw_writel(v, pll1_clk.pll_data->base + PLLOCSEL);
+ *    v = __raw_readl(pll1_clk.pll_data->base + PLLOCSEL);
+ *    pr_warning("After write %s: 0x%x\n", "PLLOCSEL", v);
+ *
+ *    v = __raw_readl(pll1_clk.pll_data->base + PLLDIV3);
+ *    pr_warning("Before write %s: 0x%x\n", "PLLDIV3", v);
+ *    v &= ~0xff;
+ *    v |= 0x8000 | 0x001F;
+ *    __raw_writel(v, pll1_clk.pll_data->base + PLLDIV3);
+ *    v = __raw_readl(pll1_clk.pll_data->base + PLLDIV3);
+ *    pr_warning("After write %s: 0x%x\n", "PLLDIV3", v);
+ *
+ *    v = __raw_readl(pll1_clk.pll_data->base + OSCDIV);
+ *    pr_warning("Before write %s: 0x%x\n", "OSCDIV", v);
+ *    v &= ~0xff;
+ *    v |= 0x8000 | 0x001F;
+ *    __raw_writel(v, pll1_clk.pll_data->base + OSCDIV);
+ *    v = __raw_readl(pll1_clk.pll_data->base + OSCDIV);
+ *    pr_warning("After write %s: 0x%x\n", "OSCDIV", v);
+ *
+ *  } else {
+ *    pr_warning("Cannot setup output clock for pll1!\n");
+ *  }
+ */
+}
+
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/devices-da8xx.c linux-3.1.10_camera/arch/arm/mach-davinci/devices-da8xx.c
--- linux-3.1.10/arch/arm/mach-davinci/devices-da8xx.c	2012-05-10 02:35:00.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/devices-da8xx.c	2014-03-21 14:28:10.413493403 +0400
@@ -553,7 +553,7 @@
 };
 
 struct da8xx_lcdc_platform_data vga_monitor_pdata = {
-	.manu_name		= "VGA monitor",
+	.manu_name		= "VGA mon",
 	.controller_data	= &lcd_cfg,
 	.type			= "VGA_Monitor",
 };
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/include/mach/da8xx.h linux-3.1.10_camera/arch/arm/mach-davinci/include/mach/da8xx.h
--- linux-3.1.10/arch/arm/mach-davinci/include/mach/da8xx.h	2012-05-11 01:32:42.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/include/mach/da8xx.h	2014-03-21 12:22:47.611260510 +0400
@@ -72,6 +72,7 @@
 
 void __init da830_init(void);
 void __init da850_init(void);
+void __init omapl138_clkout_init(void);
 
 int da830_register_edma(struct edma_rsv_info *rsv);
 int da850_register_edma(struct edma_rsv_info *rsv[2]);
diff -Nuar linux-3.1.10/arch/arm/mach-davinci/include/mach/mux.h linux-3.1.10_camera/arch/arm/mach-davinci/include/mach/mux.h
--- linux-3.1.10/arch/arm/mach-davinci/include/mach/mux.h	2012-01-18 19:33:18.000000000 +0400
+++ linux-3.1.10_camera/arch/arm/mach-davinci/include/mach/mux.h	2014-03-21 12:22:47.639260755 +0400
@@ -918,6 +918,28 @@
 	DA850_GPIO4_1,
 	DA850_GPIO6_13,
 	DA850_RTC_ALARM,
+
+	/* CLKOUT function */
+	DA850_CLKOUT,
+	DA850_GPIO2_10,
+
+	/* VPIF */
+	DA850_VPIF_DIN0,
+	DA850_VPIF_DIN1,
+	DA850_VPIF_DIN2,
+	DA850_VPIF_DIN3,
+	DA850_VPIF_DIN4,
+	DA850_VPIF_DIN5,
+	DA850_VPIF_DIN6,
+	DA850_VPIF_DIN7,
+	DA850_VPIF_DIN8,
+	DA850_VPIF_DIN9,
+	DA850_VPIF_DIN10,
+	DA850_VPIF_DIN11,
+	DA850_VPIF_DIN12,
+	DA850_VPIF_DIN13_FIELD,
+	DA850_VPIF_DIN14_HSYNC,
+	DA850_VPIF_DIN15_VSYNC
 };
 
 enum davinci_tnetv107x_index {
diff -Nuar linux-3.1.10/.config linux-3.1.10_camera/.config
--- linux-3.1.10/.config	2014-03-21 13:23:07.659160580 +0400
+++ linux-3.1.10_camera/.config	2014-03-21 14:28:26.009631011 +0400
@@ -137,6 +137,7 @@
 #
 # GCOV-based kernel profiling
 #
+# CONFIG_GCOV_KERNEL is not set
 CONFIG_HAVE_GENERIC_DMA_COHERENT=y
 CONFIG_SLABINFO=y
 CONFIG_RT_MUTEXES=y
@@ -385,24 +386,7 @@
 #
 # CPU Frequency scaling
 #
-CONFIG_CPU_FREQ=y
-CONFIG_CPU_FREQ_TABLE=y
-CONFIG_CPU_FREQ_STAT=y
-# CONFIG_CPU_FREQ_STAT_DETAILS is not set
-# CONFIG_CPU_FREQ_DEFAULT_GOV_PERFORMANCE is not set
-# CONFIG_CPU_FREQ_DEFAULT_GOV_POWERSAVE is not set
-CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE=y
-# CONFIG_CPU_FREQ_DEFAULT_GOV_ONDEMAND is not set
-# CONFIG_CPU_FREQ_DEFAULT_GOV_CONSERVATIVE is not set
-CONFIG_CPU_FREQ_GOV_PERFORMANCE=m
-CONFIG_CPU_FREQ_GOV_POWERSAVE=m
-CONFIG_CPU_FREQ_GOV_USERSPACE=y
-CONFIG_CPU_FREQ_GOV_ONDEMAND=m
-# CONFIG_CPU_FREQ_GOV_CONSERVATIVE is not set
-
-#
-# ARM CPU frequency scaling drivers
-#
+# CONFIG_CPU_FREQ is not set
 # CONFIG_CPU_IDLE is not set
 
 #
@@ -948,8 +932,8 @@
 #
 CONFIG_ARCH_REQUIRE_GPIOLIB=y
 CONFIG_GPIOLIB=y
-# CONFIG_DEBUG_GPIO is not set
-# CONFIG_GPIO_SYSFS is not set
+CONFIG_DEBUG_GPIO=y
+CONFIG_GPIO_SYSFS=y
 
 #
 # Memory mapped GPIO drivers:
@@ -1050,7 +1034,211 @@
 # CONFIG_MFD_WL1273_CORE is not set
 # CONFIG_MFD_AAT2870_CORE is not set
 # CONFIG_REGULATOR is not set
-# CONFIG_MEDIA_SUPPORT is not set
+CONFIG_MEDIA_SUPPORT=y
+
+#
+# Multimedia core support
+#
+# CONFIG_MEDIA_CONTROLLER is not set
+CONFIG_VIDEO_DEV=y
+CONFIG_VIDEO_V4L2_COMMON=y
+# CONFIG_DVB_CORE is not set
+CONFIG_VIDEO_MEDIA=y
+
+#
+# Multimedia drivers
+#
+CONFIG_RC_CORE=y
+CONFIG_LIRC=y
+CONFIG_RC_MAP=y
+CONFIG_IR_NEC_DECODER=y
+CONFIG_IR_RC5_DECODER=y
+CONFIG_IR_RC6_DECODER=y
+CONFIG_IR_JVC_DECODER=y
+CONFIG_IR_SONY_DECODER=y
+CONFIG_IR_RC5_SZ_DECODER=y
+CONFIG_IR_MCE_KBD_DECODER=y
+CONFIG_IR_LIRC_CODEC=y
+# CONFIG_IR_IMON is not set
+# CONFIG_IR_MCEUSB is not set
+# CONFIG_IR_REDRAT3 is not set
+# CONFIG_IR_STREAMZAP is not set
+# CONFIG_RC_LOOPBACK is not set
+# CONFIG_MEDIA_ATTACH is not set
+CONFIG_MEDIA_TUNER=y
+CONFIG_MEDIA_TUNER_CUSTOMISE=y
+
+#
+# Customize TV tuners
+#
+CONFIG_MEDIA_TUNER_SIMPLE=m
+CONFIG_MEDIA_TUNER_TDA8290=m
+CONFIG_MEDIA_TUNER_TDA827X=m
+CONFIG_MEDIA_TUNER_TDA18271=m
+CONFIG_MEDIA_TUNER_TDA9887=m
+CONFIG_MEDIA_TUNER_TEA5761=m
+CONFIG_MEDIA_TUNER_TEA5767=m
+CONFIG_MEDIA_TUNER_MT20XX=m
+CONFIG_MEDIA_TUNER_MT2060=m
+CONFIG_MEDIA_TUNER_MT2266=m
+CONFIG_MEDIA_TUNER_MT2131=m
+CONFIG_MEDIA_TUNER_QT1010=m
+CONFIG_MEDIA_TUNER_XC2028=m
+CONFIG_MEDIA_TUNER_XC5000=m
+CONFIG_MEDIA_TUNER_XC4000=m
+CONFIG_MEDIA_TUNER_MXL5005S=m
+CONFIG_MEDIA_TUNER_MXL5007T=m
+CONFIG_MEDIA_TUNER_MC44S803=m
+CONFIG_MEDIA_TUNER_MAX2165=m
+CONFIG_MEDIA_TUNER_TDA18218=m
+CONFIG_MEDIA_TUNER_TDA18212=m
+CONFIG_VIDEO_V4L2=y
+CONFIG_VIDEOBUF_GEN=y
+CONFIG_VIDEOBUF_DMA_CONTIG=y
+CONFIG_VIDEOBUF2_CORE=y
+CONFIG_VIDEO_CAPTURE_DRIVERS=y
+CONFIG_VIDEO_ADV_DEBUG=y
+# CONFIG_VIDEO_FIXED_MINOR_RANGES is not set
+CONFIG_VIDEO_HELPER_CHIPS_AUTO=y
+CONFIG_VIDEO_IR_I2C=y
+
+#
+# Audio decoders, processors and mixers
+#
+
+#
+# RDS decoders
+#
+
+#
+# Video decoders
+#
+
+#
+# Video and audio decoders
+#
+
+#
+# MPEG video encoders
+#
+
+#
+# Video encoders
+#
+
+#
+# Camera sensor devices
+#
+
+#
+# Flash devices
+#
+
+#
+# Video improvement chips
+#
+
+#
+# Miscelaneous helper chips
+#
+CONFIG_VIDEO_VPSS_SYSTEM=y
+CONFIG_VIDEO_VPFE_CAPTURE=y
+# CONFIG_VIDEO_DM6446_CCDC is not set
+CONFIG_VIDEO_CPIA2=y
+# CONFIG_VIDEO_SR030PC30 is not set
+# CONFIG_VIDEO_NOON010PC30 is not set
+CONFIG_SOC_CAMERA=y
+# CONFIG_SOC_CAMERA_IMX074 is not set
+# CONFIG_SOC_CAMERA_MT9M001 is not set
+# CONFIG_SOC_CAMERA_MT9M111 is not set
+# CONFIG_SOC_CAMERA_MT9T031 is not set
+# CONFIG_SOC_CAMERA_MT9T112 is not set
+# CONFIG_SOC_CAMERA_MT9V022 is not set
+# CONFIG_SOC_CAMERA_RJ54N1 is not set
+# CONFIG_SOC_CAMERA_TW9910 is not set
+CONFIG_SOC_CAMERA_PLATFORM=y
+# CONFIG_SOC_CAMERA_OV2640 is not set
+CONFIG_SOC_CAMERA_OV5642=y
+# CONFIG_SOC_CAMERA_OV6650 is not set
+# CONFIG_SOC_CAMERA_OV772X is not set
+# CONFIG_SOC_CAMERA_OV9640 is not set
+# CONFIG_SOC_CAMERA_OV9740 is not set
+# CONFIG_VIDEO_SH_MOBILE_CSI2 is not set
+# CONFIG_VIDEO_SH_MOBILE_CEU is not set
+CONFIG_V4L_USB_DRIVERS=y
+# CONFIG_USB_VIDEO_CLASS is not set
+CONFIG_USB_VIDEO_CLASS_INPUT_EVDEV=y
+CONFIG_USB_GSPCA=m
+# CONFIG_USB_M5602 is not set
+# CONFIG_USB_STV06XX is not set
+# CONFIG_USB_GL860 is not set
+# CONFIG_USB_GSPCA_BENQ is not set
+# CONFIG_USB_GSPCA_CONEX is not set
+# CONFIG_USB_GSPCA_CPIA1 is not set
+# CONFIG_USB_GSPCA_ETOMS is not set
+# CONFIG_USB_GSPCA_FINEPIX is not set
+# CONFIG_USB_GSPCA_JEILINJ is not set
+# CONFIG_USB_GSPCA_KINECT is not set
+# CONFIG_USB_GSPCA_KONICA is not set
+# CONFIG_USB_GSPCA_MARS is not set
+# CONFIG_USB_GSPCA_MR97310A is not set
+# CONFIG_USB_GSPCA_NW80X is not set
+# CONFIG_USB_GSPCA_OV519 is not set
+# CONFIG_USB_GSPCA_OV534 is not set
+# CONFIG_USB_GSPCA_OV534_9 is not set
+# CONFIG_USB_GSPCA_PAC207 is not set
+# CONFIG_USB_GSPCA_PAC7302 is not set
+# CONFIG_USB_GSPCA_PAC7311 is not set
+# CONFIG_USB_GSPCA_SE401 is not set
+# CONFIG_USB_GSPCA_SN9C2028 is not set
+# CONFIG_USB_GSPCA_SN9C20X is not set
+# CONFIG_USB_GSPCA_SONIXB is not set
+# CONFIG_USB_GSPCA_SONIXJ is not set
+# CONFIG_USB_GSPCA_SPCA500 is not set
+# CONFIG_USB_GSPCA_SPCA501 is not set
+# CONFIG_USB_GSPCA_SPCA505 is not set
+# CONFIG_USB_GSPCA_SPCA506 is not set
+# CONFIG_USB_GSPCA_SPCA508 is not set
+# CONFIG_USB_GSPCA_SPCA561 is not set
+# CONFIG_USB_GSPCA_SPCA1528 is not set
+# CONFIG_USB_GSPCA_SQ905 is not set
+# CONFIG_USB_GSPCA_SQ905C is not set
+# CONFIG_USB_GSPCA_SQ930X is not set
+# CONFIG_USB_GSPCA_STK014 is not set
+# CONFIG_USB_GSPCA_STV0680 is not set
+# CONFIG_USB_GSPCA_SUNPLUS is not set
+# CONFIG_USB_GSPCA_T613 is not set
+# CONFIG_USB_GSPCA_TV8532 is not set
+# CONFIG_USB_GSPCA_VC032X is not set
+# CONFIG_USB_GSPCA_VICAM is not set
+# CONFIG_USB_GSPCA_XIRLINK_CIT is not set
+# CONFIG_USB_GSPCA_ZC3XX is not set
+# CONFIG_VIDEO_PVRUSB2 is not set
+# CONFIG_VIDEO_HDPVR is not set
+# CONFIG_VIDEO_EM28XX is not set
+# CONFIG_VIDEO_CX231XX is not set
+# CONFIG_VIDEO_USBVISION is not set
+# CONFIG_USB_ET61X251 is not set
+# CONFIG_USB_SN9C102 is not set
+# CONFIG_USB_PWC is not set
+# CONFIG_USB_ZR364XX is not set
+# CONFIG_USB_STKWEBCAM is not set
+# CONFIG_USB_S2255 is not set
+# CONFIG_V4L_MEM2MEM_DRIVERS is not set
+CONFIG_RADIO_ADAPTERS=y
+# CONFIG_I2C_SI4713 is not set
+# CONFIG_RADIO_SI4713 is not set
+# CONFIG_USB_DSBR is not set
+# CONFIG_RADIO_SI470X is not set
+# CONFIG_USB_MR800 is not set
+# CONFIG_RADIO_TEA5764 is not set
+# CONFIG_RADIO_SAA7706H is not set
+# CONFIG_RADIO_TEF6862 is not set
+# CONFIG_RADIO_WL1273 is not set
+
+#
+# Texas Instruments WL128x FM driver (ST based)
+#
 
 #
 # Graphics support
@@ -1144,7 +1332,61 @@
 # CONFIG_SND_SOC_ALL_CODECS is not set
 CONFIG_SND_SOC_TLV320AIC3X=y
 # CONFIG_SOUND_PRIME is not set
-# CONFIG_HID_SUPPORT is not set
+CONFIG_HID_SUPPORT=y
+CONFIG_HID=y
+CONFIG_HIDRAW=y
+
+#
+# USB Input Devices
+#
+CONFIG_USB_HID=y
+CONFIG_HID_PID=y
+CONFIG_USB_HIDDEV=y
+
+#
+# Special HID drivers
+#
+# CONFIG_HID_A4TECH is not set
+# CONFIG_HID_ACRUX is not set
+# CONFIG_HID_APPLE is not set
+# CONFIG_HID_BELKIN is not set
+# CONFIG_HID_CHERRY is not set
+# CONFIG_HID_CHICONY is not set
+# CONFIG_HID_PRODIKEYS is not set
+# CONFIG_HID_CYPRESS is not set
+# CONFIG_HID_DRAGONRISE is not set
+# CONFIG_HID_EMS_FF is not set
+# CONFIG_HID_EZKEY is not set
+# CONFIG_HID_HOLTEK is not set
+# CONFIG_HID_KEYTOUCH is not set
+# CONFIG_HID_KYE is not set
+# CONFIG_HID_UCLOGIC is not set
+# CONFIG_HID_WALTOP is not set
+# CONFIG_HID_GYRATION is not set
+# CONFIG_HID_TWINHAN is not set
+# CONFIG_HID_KENSINGTON is not set
+# CONFIG_HID_LCPOWER is not set
+# CONFIG_HID_LOGITECH is not set
+# CONFIG_HID_MICROSOFT is not set
+# CONFIG_HID_MONTEREY is not set
+# CONFIG_HID_MULTITOUCH is not set
+# CONFIG_HID_NTRIG is not set
+# CONFIG_HID_ORTEK is not set
+# CONFIG_HID_PANTHERLORD is not set
+# CONFIG_HID_PETALYNX is not set
+# CONFIG_HID_PICOLCD is not set
+# CONFIG_HID_QUANTA is not set
+# CONFIG_HID_ROCCAT is not set
+# CONFIG_HID_SAMSUNG is not set
+# CONFIG_HID_SONY is not set
+# CONFIG_HID_SPEEDLINK is not set
+# CONFIG_HID_SUNPLUS is not set
+# CONFIG_HID_GREENASIA is not set
+# CONFIG_HID_SMARTJOYPLUS is not set
+# CONFIG_HID_TOPSEED is not set
+# CONFIG_HID_THRUSTMASTER is not set
+# CONFIG_HID_ZEROPLUS is not set
+# CONFIG_HID_ZYDACRON is not set
 CONFIG_USB_SUPPORT=y
 CONFIG_USB_ARCH_HAS_HCD=y
 CONFIG_USB_ARCH_HAS_OHCI=y
@@ -1371,6 +1613,7 @@
 # CONFIG_EXT3_FS_SECURITY is not set
 # CONFIG_EXT4_FS is not set
 CONFIG_JBD=y
+# CONFIG_JBD_DEBUG is not set
 CONFIG_FS_MBCACHE=y
 # CONFIG_REISERFS_FS is not set
 # CONFIG_JFS_FS is not set
@@ -1532,7 +1775,7 @@
 #
 # Kernel hacking
 #
-# CONFIG_PRINTK_TIME is not set
+CONFIG_PRINTK_TIME=y
 CONFIG_DEFAULT_MESSAGE_LOGLEVEL=4
 CONFIG_ENABLE_WARN_DEPRECATED=y
 CONFIG_ENABLE_MUST_CHECK=y
@@ -1540,7 +1783,7 @@
 # CONFIG_MAGIC_SYSRQ is not set
 # CONFIG_STRIP_ASM_SYMS is not set
 # CONFIG_UNUSED_SYMBOLS is not set
-# CONFIG_DEBUG_FS is not set
+CONFIG_DEBUG_FS=y
 # CONFIG_HEADERS_CHECK is not set
 # CONFIG_DEBUG_SECTION_MISMATCH is not set
 CONFIG_DEBUG_KERNEL=y
@@ -1585,6 +1828,7 @@
 # CONFIG_BACKTRACE_SELF_TEST is not set
 # CONFIG_DEBUG_BLOCK_EXT_DEVT is not set
 # CONFIG_DEBUG_FORCE_WEAK_PER_CPU is not set
+# CONFIG_LKDTM is not set
 # CONFIG_FAULT_INJECTION is not set
 # CONFIG_LATENCYTOP is not set
 # CONFIG_SYSCTL_SYSCALL_CHECK is not set
@@ -1606,6 +1850,7 @@
 # CONFIG_PROFILE_ALL_BRANCHES is not set
 # CONFIG_STACK_TRACER is not set
 # CONFIG_BLK_DEV_IO_TRACE is not set
+# CONFIG_DYNAMIC_DEBUG is not set
 # CONFIG_DMA_API_DEBUG is not set
 # CONFIG_ATOMIC64_SELFTEST is not set
 # CONFIG_SAMPLES is not set
